doc/design/dynamic_tls.md

```diff
@@ -0,0 +1,66 @@
+# Dynamic TLS Certs for etcd Cluster in etcd Operator
+
+Author: Eric Chiang(eric.chiang@coreos.com), Hongchao Deng(hongchao.deng@coreos.com)
+
+Motivation:
+- Simplify TLS cert generation for etcd cluster.
+- Self-hosted etcd wants to depend on Node IP, not hostnames.
+
+Requirements:
+- Dynamically generate certs per etcd pod
+- Able to use host IP in self hosted.
+
+### Option 1: K8S CSR, generate on etcd pod init container
+
+[Kubernetes certificates API][k8s_csr] provides the ability to send and approve CSR.
+
+Workflow:
+- etcd operator creates an etcd pod.
+  There is a separate service account for etcd pod to access  CSR API.
+- etcd pod first runs an init container to send a CSR and then wait for approval.
+- An external process should verify the CSR and approve it if correct.
+- Once approved, etcd pod gets the certs and bootstraps.
+
+Pros:
+- Hooks into existing k8s.
+- Delegates CA signing cert management. Not touching the signing certs and keys directly.
+
+Cons:
+- Share the same CA signing cert + key as k8s.
+- Requires etcd to be trusted by k8s cluster CA (probably only appropriate for self-hosted etcd).
```

Comment:
Yeah, this is only appropriate for self-hosting. You don't want your app etcd db to be able to access the k8s main etcd.